### PR TITLE
Fine-grain inference feedback

### PR DESCRIPTION
**Pull request by danforbes**
Related to #151 and #205

### CODE REVIEW

**Comment by danforbes** [comment]
Just realized I never actually implemented `Halt` on the `InferenceSession` side 😅

**Comment by danforbes** [comment]
I'm not able to see why the job is failing
image

**Comment by philpax** [comment]
This looks good to me from a glance, but I'll need to pull it down locally and play around with it. Will try to get to it soon 🚀

File: crates/llm-base/src/util.rs [hunk 1]
```diff
@@ -60,14 +62,15 @@ impl TokenUtf8Buffer {
...
     pub fn adapt_callback<'a, E: std::error::Error + 'static>(
-        mut callback: impl FnMut(&str) -> Result<(), E> + 'a,
-    ) -> impl FnMut(&[u8]) -> Result<(), E> + 'a {
+        mut callback: impl FnMut(InferenceResponse) -> Result<InferenceFeedback, E> + 'a,
+    ) -> impl FnMut(&[u8]) -> Result<crate::inference_session::InferenceFeedback, E> + 'a {
         let mut buffer = Self::new();
         move |token| match buffer.push(token) {
-            Some(tokens) => callback(&tokens),
-            None => Ok(()),
+            Some(tokens) => callback(InferenceResponse::PromptToken(tokens)),
```
Comment (philpax) [inline crates/llm-base/src/util.rs:72]: Hm, might be a little counterintuitive for it to always be a `PromptToken`. I'd suggest adding that to the name of the function (it seems like we only ever use it in a prompt context) or adding a parameter to switch between prompt and inference tokens.

**Review by philpax — APPROVED** [review]
LGTM, apart from that one quibble.

File: crates/llm-base/src/util.rs (hunk 1 above)
Comment (danforbes) [reply crates/llm-base/src/util.rs]: Okay, I just pulled it into `inference_session.rs` and gave it a more appropriate name :+1:

**Pull request merged**